docs: updating active configs in s3 (#1420)
CDCgov/dibbs-ecr-refiner: docs/decisions/0011_2026-06-24_updating-active-configs-in-s3.md

It would also still need access to Postgres, S3, and the same activation logic already used by the app.

## Decision Outcome

Use the ops container to run an active configuration migration whenever the `active.json` schema changes.

The migration will treat `active.json` as a generated artifact and rebuild it from the configuration data in Postgres. It will not transform the existing S3 JSON directly.

### Add explicit schema versioning

Every newly generated `active.json` file should declare the schema it uses:

```json
{
  "schema_version": 2,
  "sections": [],
  "included_condition_rsg_codes": [],
  "code_system_sets": {}
}
```

This version is only for the shape of `active.json`. It is separate from the existing configuration version used in the S3 path and `current.json`.

Lambda should validate this value after loading the file. If the version is not supported, it should raise and log a clear compatibility error instead of only returning a Pydantic validation error.

```py
if data.get("schema_version") != CURRENT_ACTIVE_CONFIG_SCHEMA_VERSION:
    raise IncompatibleActiveConfigurationError(...)

> ```

Reviewer: would love to also see some of the discussion points we had about using the serialization function in an integration test (perhaps asserting against an artifact of some sort), and having the dev in question bumping the schema version whenever there's a failure on that test.

### Reuse the existing activation functions

The activation endpoint already uses the functions needed to rebuild the S3 artifacts:

```py
convert_config_to_storage_payload()
get_config_payload_metadata()
upload_configuration_payload()
```

The ops migration should import and call these functions directly. It should not call the activation endpoint over HTTP.

The migration should not call:

```py
activate_configuration_db()
upload_condition_mapping_payload()
upload_current_version_file()
```

The configuration is already active, so the migration should not change its status, activation history, configuration version, condition mapping, or `current.json`.

### Migration flow

> 

Reviewer 2: I think we should have a note or a paragraph about needing to address the race condition where the new Lambda is deployed before the ops container finishes migrating S3. I think something along the lines as being explicit that we're going to leverage SQS visibility timeouts/retries. That way any Lambda hits to `IncompatibleActiveConfigurationError` during the migration window will raise a transient error and let the message return to the queue to be retried.

Reviewer 2: Maybe something like this?

```suggestion

To avoid deployment locks, the system will use SQS retries as a natural buffer. Lambda will be updated to raise transient errors on schema mismatch, allowing eCRs to be retried until the ops container completes the S3 migration.

```

> ```text
> Ops container starts
>     ↓
> Database migrations run
>     ↓
> TES data is loaded

Reviewer 2: Should we have an additional step here? I'm thinking something to avoid unnecessary S3 writes and log noise. I think we should introduce a `SCHEMA_VERSION` constant in the code and a tracking table in Postgres.

```suggestion
```text
Ops container starts
    ↓
Query Postgres migration table for `current_schema_version`. If there's a match with `CODE_SCHEMA_VERSION`, skip the migration.
    ↓
Database migrations run
    ↓
TES data is loaded
```

Reviewer: Worth calling out here also that the application and ops containers run at the same time, so we'll need to ensure the implementation doesn't rely on them finishing sequentially.

    ↓
Query active configurations from Postgres
    ↓
Rebuild each active.json using current code
    ↓
Overwrite the existing active.json in S3
    ↓
Application starts
```

For each active configuration, the migration should:

1. Load the active configuration from Postgres.
2. Build the current payload with `convert_config_to_storage_payload()`.
3. Build the metadata with `get_config_payload_metadata()`.
4. Upload the files with `upload_configuration_payload()`.
5. Log whether the rebuild succeeded or failed.

The regenerated payload should be written to the same S3 key:

```text
configurations/{jurisdiction}/{condition_uuid}/{configuration_version}/active.json
```

The existing file does not need to be deleted first. Writing to the same S3 key replaces the object.

The existing configuration version and `current.json` stay unchanged. Lambda continues loading the same path it uses today.

### Logging and observability

> The migration should log:
> 
> - configuration ID
> - configuration version
> - jurisdiction
> - condition UUID
> - S3 key
> - schema version written
> - migration success or failure
> 
> Lambda should log similar information when it finds a missing, invalid, or incompatible active file, including the expected and actual schema versions.
> 
> This gives the team visibility if a migration is missed or fails in production.

Reviewer 2: I think we need to clarify this section a bit more around the failure strategy. To prevent a single corrupt configuration from blocking the entire deployment pipeline, let's be explicit about a 'partial success' approach.

```suggestion
Lambda should log similar information when it finds a missing, invalid, or incompatible active file, including the expected and actual schema versions.

If an individual configuration fails to rebuild, the migration will log a critical error for that specific ID but continue with the remaining configurations. The global migration status in the DB will be marked as `PARTIAL_FAILURE` to trigger an alert for manual intervention.

This gives the team visibility if a migration is missed or fails in production.
```

#### `feat: add active configuration regeneration command`

Add an ops command that queries active configurations and regenerates their existing `active.json` files in S3.

#### `feat: run active payload migration during startup`

Run the active payload migration after database migrations and TES data loading are complete.

#### `chore: add ops permissions for configuration storage`

Give the ops container the permissions required to read and overwrite active configuration artifacts in S3.

#### `test: cover active payload regeneration`

Verify that regeneration updates the S3 payload without changing the configuration version, status, activation history, condition mapping, or `current.json`.

#### `docs: document active payload migrations`

Document the difference between configuration versions and active payload schema versions and describe the process for future schema changes.

### Open Questions and Next Steps

- Enumerate possible race conditions when the ops migration runs while Lambda is processing eCRs.
- Confirm whether the ops container can read and write Postgres.

> - Confirm whether the ops container can read and write S3 configuration artifacts.

Reviewer 2: I think we have answers for these now right? The infra team confirmed that the ops container can be provisioned with the required IAM roles for S3 `PutObject`/`GetObject` and Postgres access.
